drivers/sdhc/sdhci_common.c

```diff
@@ -0,0 +1,794 @@
+	clk_reg |= FIELD_GET(SDHCI_DIV_HI_MASK, divider) << SDHCI_DIVIDER_HI_SHIFT;
+	clk_reg |= SDHCI_CLOCK_INT_EN;
+	LOG_DBG("sdhc:clk_set: SDHCI_CLOCK_CONTROL:%04x divider:%u", clk_reg, divider);
+	sys_write16(clk_reg, sdhci_ctx->reg_base + SDHCI_CLOCK_CONTROL);
+
+
+	timeout = SDHCI_CLK_STABLE_TIMEOUT_MS * USEC_PER_MSEC;
+	/* wait for Internal Clock Stable */
+	if (!WAIT_FOR(((sys_read16(sdhci_ctx->reg_base + SDHCI_CLOCK_CONTROL) &
+			SDHCI_CLOCK_INT_STABLE)),
+		      timeout, k_msleep(1))) {
+		LOG_ERR("sdhc:clk_set: clk stable timeout");
+		return -ETIMEDOUT;
+	}
+
+	clk_reg |= SDHCI_CLOCK_CARD_EN;
+	LOG_DBG("sdhc:clk_set: stable SDHCI_CLOCK_CONTROL:%04x", clk_reg);
+	sys_write16(clk_reg, sdhci_ctx->reg_base + SDHCI_CLOCK_CONTROL);
+
+	return 0;
+}
+
+int sdhci_set_bus_width(struct sdhci_common *sdhci_ctx, enum sdhc_bus_width bus_width)
+{
+	uint8_t ctrl;
+
+	/* Set bus width */
+	ctrl = sys_read8(sdhci_ctx->reg_base + SDHCI_HOST_CONTROL);
+
+	ctrl &= ~SDHCI_CTRL_8BITBUS;
+
+	switch (bus_width) {
+	case SDHC_BUS_WIDTH1BIT:
+		ctrl &= ~SDHCI_CTRL_4BITBUS;
+		break;
+	case SDHC_BUS_WIDTH4BIT:
+		ctrl |= SDHCI_CTRL_4BITBUS;
+		break;
+	case SDHC_BUS_WIDTH8BIT:
+		ctrl |= SDHCI_CTRL_8BITBUS;
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	LOG_DBG("sdhc:bus_width: SDHCI_HOST_CONTROL:%02x", ctrl);
+	sys_write8(ctrl, sdhci_ctx->reg_base + SDHCI_HOST_CONTROL);
+	return 0;
+}
+
+int sdhci_set_voltage(struct sdhci_common *sdhci_ctx, enum sd_voltage signal_voltage)
+{
+	uint16_t ctrl;
+
+	ctrl = sys_read16(sdhci_ctx->reg_base + SDHCI_HOST_CONTROL2);
+
+	switch (signal_voltage) {
+	case SD_VOL_3_3_V:
+		ctrl &= ~SDHCI_CTRL_VDD_180;
+		sys_write16(ctrl, sdhci_ctx->reg_base + SDHCI_HOST_CONTROL2);
+
+		/* Wait for 5ms */
+		k_msleep(5);
+
+		/* 3.3V regulator output should be stable within 5 ms */
+		ctrl = sys_read16(sdhci_ctx->reg_base + SDHCI_HOST_CONTROL2);
+		LOG_DBG("sdhc:set_3_3: SDHCI_HOST_CONTROL2: %04x", ctrl);
+		if (ctrl & SDHCI_CTRL_VDD_180) {
+			LOG_ERR("sdhc: 3.3V regulator output did not become stable\n");
+			return -EIO;
+		}
+
+		break;
+	case SD_VOL_1_8_V:
+		ctrl |= SDHCI_CTRL_VDD_180;
+		sys_write16(ctrl, sdhci_ctx->reg_base + SDHCI_HOST_CONTROL2);
+
+		/* Wait for 5 ms */
+		k_msleep(5);
+
+		/* 1.8V regulator output has to be stable within 5 ms */
+		ctrl = sys_read16(sdhci_ctx->reg_base + SDHCI_HOST_CONTROL2);
+		LOG_DBG("sdhc:set_1_8: SDHCI_HOST_CONTROL2: %04x", ctrl);
+		if (!(ctrl & SDHCI_CTRL_VDD_180)) {
+			LOG_ERR("sdhc: 1.8V regulator output did not become stable\n");
+			return -EIO;
+		}
+
+		break;
+	default:
+		LOG_DBG("sdhc: signal voltage not supported %d", signal_voltage);
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+int sdhci_set_uhs_timing(struct sdhci_common *sdhci_ctx, enum sdhc_timing_mode timing)
+{
+	bool hspd = false;
+	uint16_t reg;
+
+	reg = sys_read16(sdhci_ctx->reg_base + SDHCI_HOST_CONTROL2);
+	reg &= ~SDHCI_CTRL_UHS_MASK;
+
+	switch (timing) {
+	case SDHC_TIMING_LEGACY:
```

Comment:
you trying use UHS-I timing bits for setting `legacy mode`. Pls, review Host Control 2 register description (2.2.24 Host Control 2 Register (Offset 03Eh)) from `SD Host Controller Simplified Specification Version 3.00`